### ISSUE TITLE
Reproducing Results on SVG (BAIR dataset)

### ISSUE THREAD

**Issue by kaanakan** [edouardelasalles/srvp#6]
Hi,

Thank you for providing the code of the work.

In the paper,  it is stated that 

"for each test sequence, sampling from the tested model a given number (here, 100) of
possible futures and reporting the best performing sample
against the true video"

What I understand from this is that generate hundred samples for each test sequence and report the best SSIM value for each test sequence. Is this true?

In SVG evaluation, there is no metric report but the metrics are calculated. For BAIR dataset, the metrics for SSIM and PSNR is higher than you report on the paper (Table 5.). In the table, it is reported as 18.95 ± 0.26 (PSNR), 0.8058 ± 0.0088 (SSIM), however, if we take the best performing sequence from 100 samples, metrics jump to 22.70 (PSNR), 0.8910 (SSIM). (Weights from SVG repository are used.) Am I doing something in this evaluation?

If you clarify these points, I would be grateful.
Thanks.

**Comment by edouardelasalles** [comment]
Hi, thank you for your interest in our work!

Indeed, we report for PSNR, SSIM and LPIPS their best value with respect to the ground truth over 100 samples from the model.

We also used the weights from the pretrained models of SVG, so we should obtain the same performances. Given the results that you mentioned, you might be computing them on a different sequence length, or you may have included the conditioning frames in your computation.
On the BAIR dataset, we use 2 conditioning frames for test sequences of size 30. Therefore, the results we provide are computed on the last 28 frames.

**Comment by kaanakan** [comment]
Hi,

for clarification, I will explain all procedure for getting results. First of all,  I am using the paper values which are 2 conditioning frames and 28 future frames.

I used batch size of 64 for evaluation.
[after this line](https://github.com/edenton/svg/blob/3f19f0b581161614382b2d529f8d92c7d25999e5/generate_svg_lp.py#L172 )
I saved the ssim and psnr arrays which have the shape of (64, 100, 28). In total, we have 4 of these arrays. At the end, I concatenate the arrays and we have (256, 100, 28). 

In this part, if we want to take the best value for each testing sequence, we need to take the maximum along the axis 1.
Here I calculate the best samples for each testing sequence,
`
best_samples = np.amax(ssim_array, 1), where ssim_array is (256,100,28)
`
`
ssim_score = np.mean(best_samples)
`

Am I missing something? 
Thanks a lot!

**Comment by edouardelasalles** [comment]
By doing that you actually take the best frame at each time step among the 100 samples. Hence, your result is composed of 28 frames that do not necessarily come from the same sequence. What you have to do instead is take the best **sequence**, that is the complete test sequence that yields the best performance.

**Comment by kaanakan** [comment]
Oh, I understand now. If I implement it correctly, again, I could not get your results. I used,

`
ssim_of_sequences = np.mean(ssim_values, 2)
`
`
ssim_of_best_sequences = np.max(ssim_of_sequences, 1)
`
`
result = np.mean(ssim_of_best_sequences)
`

ssim_values -> [256, 100, 28] array, each entry is ssim value
ssim_of_sequences -> [256, 100] array, each entry is the mean ssim value of a sequence
ssim_of_best_sequences -> [256] array, each entry is the ssim value of the best sequence
result value is 0.8454 for ssim and the mean of all frames (mean(ssim_values)) is 0.7949

Am I missing something?
Thanks a lot!

**Comment by edouardelasalles** [comment]
Does the PSNR now match the results that we report?

The discrepancy could come from the differences between the SSIM implementation of the SVG repository and ours. You can try to compute SSIM with [the function we provide](https://github.com/edouardelasalles/srvp/blob/adbf86f4c4a348853bccb44492190ee7c47500e3/test.py#L36).

More particularly, in SVG's implementation, SSIM is computed with respect to a data range of [-1, 1], whereas ours is computed with respect to a data range of [0, 1] like the input images, which changes the SSIM values.

**Comment by kaanakan** [comment]
I was using a different dataset, I changed the dataset to your version.
I still use the calculations in SVG repository.


> Does the PSNR now match the results that we report?

No, it is 19.56. They used skimage's PSNR which calculates in the way you did, I think.


> More particularly, in SVG's implementation, SSIM is computed with respect to a data range of [-1, 1], whereas ours is computed with respect to a data range of [0, 1] like the input images, which changes the SSIM values.

Are you sure about this part? I think, in SVG, SSIM is computed with respect to a data range of [0, 1] by using skimage.metrics' structural similarity  function. 


I am not sure about the problem. It maybe because of the python version because SVG repo doesn't give an environment information. Can you give any suggestions?

Thanks.

**Comment by edouardelasalles** [comment]
## SSIM

You can see [here](https://github.com/edenton/svg/blob/master/utils.py#L217) that no data range value is passed to skimage's SSIM function in SVG's code. However, we can see [here](https://github.com/scikit-image/scikit-image/blob/v0.17.2/skimage/metrics/_structural_similarity.py#L159) that if no data range is passed it is then inferred from the data type. In our and SVG's case, these variables are of type float, and the resulting `dmin` and `dmax` values will be `-1` and `+1` respectively (c.f. [here](https://github.com/scikit-image/scikit-image/blob/master/skimage/util/dtype.py#L25)). Thus, in the SVG repository, SSIM is computed with a data range of 2, whereas in our case it is computed with a data range of 1, which is closest to the data, and accordingly alters SSIM values. You should obtain results similar to those that we report by setting the data range to 1 in SVG's test code.

## PSNR

We performed further tests and found out that there may be some variance on SVG's results when performing the test protocol several times. Therefore, we think that your PSNR results are coherent with ours. Note that our results analysis remains unchanged.

**Comment by kaanakan** [comment]
Thank you for help.

I have changed the data range parameter of metrics calculations and the results were different again. I suppose it is because of the version differences in SVG repository (since there is no requirements file, I have used different pytorch versions). 

The results I got after changing data range to 1 was 19.56 (again) for PSNR, and about 0.82 for SSIM. I think the results will probably be the same with yours if I save all the images and use your code to calculate metrics.

Not sure to close the issue, feel free to close it.

Thank you for help. 
Regards.

**Comment by edouardelasalles** [comment]
You're welcome. I will then close the issue. Feel free to re-open it if needed.

**Comment by kaanakan** [comment]
Hi, 

I am re-writing because I want to ask how did you manage to get the results of SVG on datasets. Can you provide a little bit detail for the process? For example, which pytorch version did you use, did you use pre-trained weights they provided or train a new model? 

I would be really appreciated if you can answers these questions.
Thanks a lot.

**Comment by edouardelasalles** [comment]
Hi,
Sorry for the late reply, we did not see your question as the issue remained closed. Feel free to re-open it if you have any other question.
To reproduce SVG results, we used PyTorch 0.4.0. As specified in the end of page 5 and in Appendix B of the paper, we used the code and pre-trained weights provided in SVG's official GitHub repository. The only exception is for Moving MNIST, where we trained the model from scratch with the provided code on our version of the dataset.
